pyxdsm/XDSM.py

```diff
@@ -19,744 +23,536 @@
-        if spec_name is None:
-            spec_name = node_name
-
-        sys = System(node_name, style, label, stack, faded, label_width, spec_name)
-        self.systems.append(sys)
+        # Only process if these aren't already in data (from deserialization)
+        if "optional_packages" not in data:
+            # Process optional packages
+            packages = []
+            if optional_latex_packages is not None:
+                if isinstance(optional_latex_packages, str):
+                    packages = [optional_latex_packages]
+                elif isinstance(optional_latex_packages, list):
+                    packages = optional_latex_packages
+                else:
+                    raise ValueError("optional_latex_packages must be a string or list of strings")
+            data["optional_packages"] = packages
+
+        if "auto_fade" not in data:
+            # Process auto_fade
+            fade_config = AutoFadeConfig()
+            if auto_fade is not None:
+                fade_config = AutoFadeConfig(**auto_fade)
+            data["auto_fade"] = fade_config
+
+        if "use_sfmath" not in data:
+            data["use_sfmath"] = use_sfmath
+
+        super().__init__(**data)
+
+    @model_validator(mode="before")
+    @classmethod
+    def _set_defaults_for_missing_fields(cls, data):
+        """Ensure missing or null collection fields get empty defaults."""
+        if not isinstance(data, dict):
+            return data
+
+        # Set empty defaults for missing or null collection fields
+        if "inputs" not in data or data.get("inputs") is None:
+            data["inputs"] = {}
+        if "outputs" not in data or data.get("outputs") is None:
+            data["outputs"] = {}
+        if "systems" not in data or data.get("systems") is None:
+            data["systems"] = []
+        if "connections" not in data or data.get("connections") is None:
+            data["connections"] = []
+        if "processes" not in data or data.get("processes") is None:
+            data["processes"] = []
+
+        return data
+
+    @model_validator(mode="after")
+    def _validate_unique_system_names(self):
+        """Ensure all system names are unique."""
+        names = [sys.node_name for sys in self.systems]
+        duplicates = [n for n in names if names.count(n) > 1]
+        if duplicates:
+            raise ValueError(f"Duplicate system names: {set(duplicates)}")
+        return self
 
-    def add_input(self, name, label, label_width=None, style="DataIO", stack=False, faded=False):
-        r"""
-        Add an input, which will appear in the top row of the diagram.
+    def add_system(
+        self,
+        node_name: str,
+        style: str,
+        label: Union[str, List[str], Tuple[str, ...]],
+        stack: bool = False,
+        faded: bool = False,
+        label_width: Optional[int] = None,
+        spec_name: Optional[str] = None,
+    ) -> None:
+        """Add a system block on the diagonal."""
```

Comment:
Add back the appropriate Parameters section of the docstrings here and below

Reply:
I've cleaned this up a bit.

I removed any changes involving MatrixEquation/Jacobian because I wasn't really interested in having that be serializable with this PR.

I've added the original docstrings back in place.

The `write_sys_specs` is the original specs method, I left it in place with the `to_json` and `from_json` methods being the pydantic interface.  Maybe there's not longer a reason for write_sys_specs but I didn't want to break any existing implementations.